test/Test/Data/List.purs

```diff
@@ -354,6 +375,13 @@ testList = do
   let xs = fromFoldable (range 1 100000)
   assert $ traverse Just xs == Just xs
 
+  log "traverseWithIndex should be stack-safe"
+  assert $ traverseWithIndex (const Just) xs == Just xs
+
+  log "traverseWithIndex should be correct"
+  assert $ traverseWithIndex (\i a -> Just $ i + a) (fromFoldable [2, 2, 2])
```

Comment:
Nitpicking, sorry: could you use an Applicative which tests that the order of effects is correct, like `Array`?

Reply:
Hmm, actually that wouldn't work, never mind.